hoodie-client/src/main/java/com/uber/hoodie/io/ConsistencyCheck.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.io;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.uber.hoodie.common.SerializableConfiguration;
+import com.uber.hoodie.common.util.FSUtils;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaSparkContext;
+
+/**
+ * Checks if all the written paths have their metadata consistent on storage and thus be listable to
+ * queries. This is important for cloud, stores like AWS S3 which are eventually consistent with
+ * their metadata. Without such checks, we may proceed to commit the written data, without the
+ * written data being made available to queries. In cases like incremental pull this can lead to
+ * downstream readers failing to ever see some data.
+ */
+public class ConsistencyCheck implements Serializable {
+
+  private static final transient Logger log = LogManager.getLogger(ConsistencyCheck.class);
+
+  private String basePath;
+
+  private List<String> relPaths;
+
+  private transient JavaSparkContext jsc;
+
+  private SerializableConfiguration hadoopConf;
+
+  private int parallelism;
+
+  public ConsistencyCheck(String basePath, List<String> relPaths, JavaSparkContext jsc,
+      int parallelism) {
+    this.basePath = basePath;
+    this.relPaths = relPaths;
+    this.jsc = jsc;
+    this.hadoopConf = new SerializableConfiguration(jsc.hadoopConfiguration());
+    this.parallelism = parallelism;
+  }
+
+  @VisibleForTesting
+  void sleepSafe(long waitMs) {
+    try {
+      Thread.sleep(waitMs);
+    } catch (InterruptedException e) {
+      // ignore & continue next attempt
+    }
+  }
+
+  /**
+   * Repeatedly lists the filesystem on the paths, with exponential backoff and marks paths found as
+   * passing the check.
+   *
+   * @return list of (relative) paths failing the check
+   */
+  public List<String> check(int maxAttempts, long initalDelayMs) {
+    long waitMs = initalDelayMs;
+    int attempt = 0;
+
+    List<String> remainingPaths = new ArrayList<>(relPaths);
+    while (attempt++ < maxAttempts) {
+      remainingPaths = jsc.parallelize(remainingPaths, parallelism)
+          .groupBy(p -> new Path(basePath, p).getParent()) // list by partition
+          .map(pair -> {
+            FileSystem fs = FSUtils.getFs(basePath, hadoopConf.get());
+            // list the partition path and obtain all file paths present
+            Set<String> fileNames = Arrays.stream(fs.listStatus(pair._1()))
```

Comment:
For a large number of relPaths, this can cause explosion of listStatus() calls on the file system. For S3, I'm not sure if that matters a lot. Anyways, don't see a way to avoid it.

Reply:
Not sure what you mean. this is listing each partition that has unverified files? Hows this any higher than what the cleaner does for example?

Reply:
This is additional, that's what I mean